Source/WebInspectorUI/UserInterface/Views/TreeOutline.js

```diff
@@ -941,49 +927,66 @@ WI.TreeOutline = class TreeOutline extends WI.Object
         }
 
         let offsetFromContainer = calculateOffsetFromContainer(this._virtualizedTopSpacer.parentNode ? this._virtualizedTopSpacer : this.element, this._virtualizedScrollContainer);
-        let numberVisible = Math.ceil(Math.max(0, this._virtualizedScrollContainer.offsetHeight - offsetFromContainer) / this._virtualizedTreeItemHeight);
-        let extraRows = Math.max(numberVisible * 5, 50);
-        let firstItem = Math.max(0, Math.floor((this._virtualizedScrollContainer.scrollTop - offsetFromContainer) / this._virtualizedTreeItemHeight) - extraRows);
-        let lastItem = firstItem + numberVisible + (extraRows * 2);
+        let scrollTop = Math.max(0, this._virtualizedScrollContainer.scrollTop - offsetFromContainer);
+        let scrollBottom = scrollTop + this._virtualizedScrollContainer.offsetHeight;
+
+        let firstVisibleIndex = Math.floor(scrollTop / this._virtualizedTreeItemHeight);
+        let lastVisibleIndex = Math.ceil(scrollBottom / this._virtualizedTreeItemHeight);
+        let numberVisible = lastVisibleIndex - firstVisibleIndex + 1;
+        if (!numberVisible)
+            return;
+
+        const skipUnrevealed = true;
+        const stayWithin = null;
+        const dontPopulate = true;
+        const ignoreHidden = true;
+
+        let firstChild = this.children[0];
+        if (!firstChild.revealed(ignoreHidden))
+            firstChild = traverseNextTreeElement.traverseNextTreeElement(skipUnrevealed, stayWithin, dontPopulate);
 
         let shouldScroll = false;
         if (focusedTreeElement && focusedTreeElement.revealed(false)) {
-            let index = walk(this, (treeElement) => treeElement === focusedTreeElement).count;
-            if (index < firstItem) {
-                firstItem = index - extraRows;
-                lastItem = index + numberVisible + extraRows;
-            } else if (index > lastItem) {
-                firstItem = index - numberVisible - extraRows;
-                lastItem = index + extraRows;
+            let index = 0;
+            for (let child = firstChild; child && child !== focusedTreeElement; child = child.traverseNextTreeElement(skipUnrevealed, stayWithin, dontPopulate))
+                ++index;
+
+            if (index < firstVisibleIndex) {
+                firstVisibleIndex = index;
+                lastVisibleIndex = index + numberVisible;
+                shouldScroll = true;
+            } else if (index > lastVisibleIndex) {
+                firstVisibleIndex = index - numberVisible;
+                lastVisibleIndex = index;
+                shouldScroll = true;
             }
-
-            // Only scroll if the `focusedTreeElement` is outside the visible items, not including
```

Comment:
So we are no longer trying to avoid requesting scroll? Or does it not do anything?

Reply:
we are.  i just moved the logic into the `if` and `else if` above to simplify the logic